go-runtime/encoding/encoding.go

```diff
@@ -107,6 +95,24 @@ func encodeValue(v reflect.Value, w *bytes.Buffer) error {
 	}
 }
 
+var ftlOptionTypePath = "github.com/TBD54566975/ftl/go-runtime/ftl.Option"
+
+func isOption(t reflect.Type) bool {
+	return strings.HasPrefix(t.PkgPath()+"."+t.Name(), ftlOptionTypePath)
+}
+
+func encodeOption(v reflect.Value, w *bytes.Buffer) error {
```

Comment:
Thanks for catching that... we originally did this to break a circular dependency between the `encoding` and `ftl` packages, but since `call`'s dependency prevents that anyway, this is now pointless. cc @worstell so we can discuss

Reply:
One option is to move the logic back to `option.go`, undo the broken encapsulation, and then special case `Option` explicitly in the `switch` in `encodeValue`.

Reply:
That sounds like a reasonable approach.

Reply:
ah actually we needed to move the implementation into encoding.go to prepare for a following change that will pass context through to the custom encoder. the json marshaling overrides for option must call into the custom encoder to marshal/unmarshal the optional’s underlying value, and we don’t have a context to pass from these callers

Reply:
a potential solution with callbacks—adding an interface like this in `encoding.go`:
```
type OptionMarshaler interface {
	Marshal(marshalSome func ... ) ([]byte, error)
	Unmarshal(unmarshalSome func ... ) error
}
```
implemented in `optional.go` (this just subs the `ftlencoding` calls from our current JSON overrides in `option.go` with the callbacks), e.g.:
```
func (o Option[T]) Marshal(marshalSome func(v any) ([]byte, error)) ([]byte, error) {
	if o.ok {
		return marshalSome(o.value) // was `ftlencoding.Marshal(o.value)`
	}
	return []byte("null"), nil
}

func (o Option[T]) Unmarshal(data []byte, unmarshalSome func(data []byte, v any) error) error {
	if string(data) == "null" {
		o.ok = false
		return nil
	}
	if err := unmarshalSome(data, &o.value); err != nil { // was `ftlencoding.Unmarshal(data, &o.value)`
		return err
	}
	o.ok = true
	return nil
}
```

then `encoding.go` uses `OptionMarshaler.Marshal/Unmarshal` if the value implements `OptionMarshaler`. the callbacks it would pass would be the original `Marshal/Unmarshal` calls